### DISCUSSION TITLE
Callback Output is not generating after task complete

### DISCUSSION THREAD

**Discussion by pip-install-skills** [crewAIInc/crewAI#1479] [Q&A]
I have created the tasks and added the python callable function, but I invoke the kickoff function after the task completion there is not txt file getting created, this is the version I'm using crewai==0.28.0

```
        def on_task_complete(task_result):
            file_path = "output.txt"
            with open(file_path, 'w') as file:
                file.write(task_result)


        task = Task(
            description=task_info.description,
            expected_output=task_info.expected_output,
            agent=agent,  # Assign the agent created on the go,
            callback=on_task_complete
        )
```
This is my code, please help

**Comment by KeepALifeUS** [comment]
This looks like the callback signature might not match what crewAI expects. Here are some things to check:

## 1. Check Callback Signature

crewAI might expect a specific signature:

```python
def on_task_complete(output):
    # output might be a TaskOutput object, not just a string
    print(f"Output type: {type(output)}")
    print(f"Output: {output}")
    
    # If it's a TaskOutput object:
    result = output.raw if hasattr(output, "raw") else str(output)
    
    with open("output.txt", "w") as f:
        f.write(result)
```

## 2. Verify Callback is Being Called

Add debugging to confirm execution:

```python
def on_task_complete(output):
    print("CALLBACK TRIGGERED")  # Should appear in console
    with open("debug.txt", "w") as f:
        f.write("Callback executed")
    # Then your actual logic
```

## 3. Alternative: Use Task State

If callbacks are unreliable, capture output via state:

```python
state = {"task_outputs": {}}

result = crew.kickoff()

# After completion, all outputs are in result
for task_output in result.tasks_output:
    state["task_outputs"][task_output.task_id] = task_output.raw
    with open(f"output_{task_output.task_id}.txt", "w") as f:
        f.write(task_output.raw)
```

## 4. Check Version Compatibility

In 0.28.0, the callback API might have changed. Check the changelog or try:

```python
@task
def my_task():
    ...
    return result  # Result goes to callback
```

Can you share what `type(task_result)` shows if you add a print statement?

**Comment by Hassanmahmood4** [comment]
A couple of things come to mind:

1. **CrewAI 0.28.0 is quite old**, and the callback behavior may differ from the current documentation. Can you confirm whether `on_task_complete` is actually being invoked? For example:

```python
def on_task_complete(task_result):
    print("Callback called!")
    print(task_result)

    with open("output.txt", "w") as f:
        f.write(str(task_result))
```

If you don't see `"Callback called!"` in the console, then the callback itself is not being executed.

2. The callback argument may not be a plain string. Try converting it explicitly:

```python
f.write(str(task_result))
```

instead of:

```python
f.write(task_result)
```

3. Check your current working directory. `output.txt` may be getting created somewhere unexpected:

```python
import os
print(os.getcwd())
```

4. If none of the above helps, it might be related to the CrewAI version. I would try upgrading to a newer version of CrewAI and testing the same code again:

```bash
pip install -U crewai
```

Could you confirm:
- Do you see the callback print statements?
- Is there any exception being raised?
- Where does `os.getcwd()` point to during execution?

That would help narrow down whether this is a callback issue, a file path issue, or a CrewAI 0.28.0 behavior difference.
